### ISSUE TITLE
Dependency options should be included in package_id.

### ISSUE THREAD

**Issue by kent-at-multiscale** [conan-io/conan#1050]
I have a library libR that depends on another library libZ. For an application that links to my library, there are 4 possible sets of options.

App -> libR.a -> libZ.a
App -> libR.so -> libZ.a
App -> libR.a -> libZ.so
App -> libR.so -> libZ.so

Right now when I build all 4 configurations, I only end up with 2 libR packages. One provides libR.a and the other provides libR.so. However, I have no way to control whether each libR needs libZ.so or whether it includes libZ.a implicitly. If I do an `objdump -x libR.so`, will this line appear or not?

```
Dynamic Section:
  NEEDED  libZ.so
```

We could fix this by making the `package_id` include the full expanded options, substituting in default options when they are not specified, and including the fully expanded options of all dependencies.

Basically, libR `package_id` would be based on `[full_options]` instead of `[options]`.

**Comment by memsharded** [comment]
You can already configure the ``package_id`` to use the dependents options if you want, with the use of the ``package_id()`` method, something like:

```python
def package_id(self):
    self.info.options = self.info.full_options
```

Or if you want to just use some of the dependencies options:
```python
def package_id(self):
    self.info.options["libZ"] = self.info.full_options["libZ"]
```

There could be another indirect way to control that, explained in: http://docs.conan.io/en/latest/howtos/define_abi_compatibility.html

If you select the ``full_package_mode``, every change of the dependency package will generate a new binary:
```python
def package_id(self):
      self.info.requires["libZ"].full_package_mode()
```

**Comment by danimtb** [comment]
Hi! This issue has been open for a while and I think it was well answered to solve the problem. Please feel free to reopen if you have any further doubts. Thanks!
